### PR TITLE
Fix incorrect message on describe netpol

### PR DESCRIPTION
**Pull request by jackkleeman**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
When I describe network policies, it often tells me that pods are
isolated for egress connectivity because the policy that applies to them
has no egress rules. However, this would only lead to isolation if there
is an explicitly set egress policy type. Otherwise, the policy allows
egress traffic. The same applies if you have explicitly set an egress
type only, describe will incorrectly report isolated ingress traffic.

This PR fixes this by inferring the applicable direction for the policy
based on the field comment on PolicyTypes, and then if a policy doesn't
apply eg to egress, we print 'Allowing all egress traffic'

Before:
```
  Allowing egress traffic:
    <none> (Selected pods are isolated for egress connectivity)
  Policy Types: Ingress
```
After:
```
  Not affecting egress traffic
  Policy Types: Ingress
```

**Does this PR introduce a user-facing change?**:

```release-note
Fix incorrect network policy description suggesting that pods are isolated when a network policy has no rules of a given type
```

### CODE REVIEW

**Comment by deads2k** [comment]
/unassign

@kubernetes/sig-network-bugs

**Comment by rikatz** [comment]
/assign

I’m no reviewer, but can take a look into it and try to help :)

**Comment by rikatz** [comment]
OK, I was going to suggest to look directly into PolicyType as it seems to return what's the policy type regardless of it existence, as the following object:

```yaml

apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: egress
  namespace: default
spec:
  podSelector:
    matchLabels:
      role: db
  egress:
  - to:
    - ipBlock:
        cidr: 10.0.0.0/24
    ports:
    - protocol: TCP
      port: 5978
```

but it seems to also insert this as an Ingress Policy, even if this isn't:

```
kubectl describe netpol egress     
Name:         egress
Namespace:    default
Created on:   2019-10-27 16:56:11 -0300 -03
Labels:       <none>
Annotations:  <none>
Spec:
  PodSelector:     role=db
  Allowing egress traffic:
    To Port: 5978/TCP
    To:
      IPBlock:
        CIDR: 10.0.0.0/24
        Except: 
  Allowing all ingress traffic
  Policy Types: Ingress, Egress
```

I need some time to recreate my environment here and check if the existence of only one Egress Policy denies all the ingress traffic (in my point of view, it shouldn't).

If it shouldn't really be, probably whatever generates this information (I would try to guess it's in the controller) this also needs to be corrected.
 
Will make further tests, but if You also can verify this behavior I would appreciate that.

After that, being some issue also in this object generation I think the best way is to correct this AND then verify directly if this is the case.

File: staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go [hunk 1]
```diff
@@ -3741,13 +3741,42 @@ func describeNetworkPolicySpec(nps networkingv1.NetworkPolicySpec, w PrefixWrite
...
+		printNetworkPolicySpecEgressTo(nps.Egress, "    ", w)
+	} else {
+		w.Write(LEVEL_1, "Allowing all egress traffic\n")
+
+	}
 	w.Write(LEVEL_1, "Policy Types: %v\n", policyTypesToString(nps.PolicyTypes))
 }
 
+func getPolicyType(nps networkingv1.NetworkPolicySpec) (bool, bool) {
+	if len(nps.PolicyTypes) == 0 {
```
Comment (rikatz) [inline staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go]: Is there any situation when creating a Network Policy this situation be true? As an example, creating a rule but not specifying its type will generate a ``nps.PolicyTypes`` with at least one object, right?

**Comment by jackkleeman** [comment]
@rikatz see field comment for PolicyTypes:
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#networkpolicyspec-v1-networking-k8s-io

File: staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go (hunk 1 above)
Comment (jackkleeman) [reply staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go]: I don't believe so! The policy type is often not set, and defaults in that case to ingress at the very least, and additionally egress if an egress rule is present

**Comment by rikatz** [comment]
OK, so creating a policy with empty PolicyType will always default to Ingress,Egress.

Otherwise it will create a Ingress or Egress Policy, so the ``if len(nps.PolicyTypes) == 0`` might never happen (or is there any situation this might happen?)

As this is the situation, maybe only the ``for`` to check which PolicyTypes the object has is enough?

File: staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go (hunk 1 above)
Comment (rikatz) [reply staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go]: So I think that only doing the ``for`` checking if this is an Ingress and/or Egress policy is enough

**Comment by rikatz** [comment]
/assign @andrewsykim 
/assign @cmluciano 

Assigning some reviewers to check if there's something else to do.

**Comment by jackkleeman** [comment]
I am not sure we are on the same page. Creating a policy with empty policy types will not cause them to be created for you. It is very common to have no policy types, and they are not backfilled by the Kubernetes API or by kubectl as far as I know. However, if they do not exist, there is a standard for inferring to what direction the policy applies without the use of policy types. This is the standard that I have implemented.

File: staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go (hunk 1 above)
Comment (jackkleeman) [reply staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go]: I don't think so

**Comment by rikatz** [comment]
> I am not sure we are on the same page. Creating a policy with empty policy types will not cause them to be created for you. It is very common to have no policy types, and they are not backfilled by the Kubernetes API or by kubectl as far as I know. However, if they do not exist, there is a standard for inferring to what direction the policy applies without the use of policy types. This is the standard that I have implemented.

@jackkleeman Sorry about being insistent, but here is my test:

```
$ cat empty.yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: empty
  namespace: default
spec:
  podSelector: {}

$ kubectl create -f empty.yaml
networkpolicy.networking.k8s.io/empty created

$ kubectl get netpol empty -o yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  creationTimestamp: "2019-10-27T23:02:02Z"
  generation: 1
  name: empty
  namespace: default
  resourceVersion: "138631"
  selfLink: /apis/networking.k8s.io/v1/namespaces/default/networkpolicies/empty
  uid: 775ef8f0-7c78-40ed-ae67-89d29cc6ad90
spec:
  podSelector: {}
  policyTypes:
  - Ingress
```

So it seems somewhere the PolicyType is backfilled with default Ingress when everything is empty. This is the object that's persisted into apiserver/etcd. This can also be verified inside the describe code here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go#L3748

**Comment by jackkleeman** [comment]
Oh! You might be right. I need to test this on my cluster as I have never seen this behaviour before.

However, I wonder what the right behaviour should be all the same. If we were being defensive, we would not assume that types were set already

**Comment by cmluciano** [comment]
I'm spinning up a cluster to test this out. I'm trying to test how describe details are working with respect to NetPol.

**Review by cmluciano — COMMENTED** [review]
Yes the comments about PolicyTypes being auto filled is correct per the implementation I created when we introduced Egress. The defaulting logic tests show what our expectations are here https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/networking/v1/defaults_test.go.

TLDR - you will always get the Ingress in that PolicyTypes map unless you create a Egress policy and explicitly set the PolicyType in your configuration. 

The defaulting decision was intentional to preserve the behavior of existing NetPol expectations that enforced Ingress policies if any network policies were created. So the only option we had for changing a v1 API was to preserve this backwards-compatible expectation but with the option to override it by explicitly filling in that PolicyType.

**Comment by jackkleeman** [comment]
I will simplify the logic, thanks

File: staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go [hunk 2]
```diff
@@ -3741,13 +3741,38 @@ func describeNetworkPolicySpec(nps networkingv1.NetworkPolicySpec, w PrefixWrite
...
+
+	}
 	w.Write(LEVEL_1, "Policy Types: %v\n", policyTypesToString(nps.PolicyTypes))
 }
 
+func getPolicyType(nps networkingv1.NetworkPolicySpec) (bool, bool) {
+	var ingress, egress bool
+	for _, pt := range nps.PolicyTypes {
+		switch pt {
+		case networkingv1.PolicyTypeIngress:
```
Comment (cmluciano) [inline staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go]: If you have a combined policy within the spec, won't this ignore the egress part?

**Review by cmluciano — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go (hunk 2 above)
Comment (jackkleeman) [reply staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go]: It will return true true, no? Maybe I'm misunderstanding you

Comment (cmluciano) [reply staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go]: For whatever reason I thought the case statement hitting the first entry would exit the loop but then I realized that this is iterating over the actual PolicyTypes within the Spec. Can you please add some unit tests the describe_tests file to verify these behaviors?

**Review by cmluciano — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go (hunk 2 above)
Comment (jackkleeman) [reply staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go]: Pushed two tests, thanks

**Comment by jackkleeman** [comment]
I'm slightly confused by this test failure as it runs locally:
```
.io/kubectl/pkg/describe/versioned
+++ [1108 17:07:30] Building go targets for darwin/amd64:
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
+++ [1108 17:08:07] Building go targets for darwin/amd64:
    ./vendor/github.com/go-bindata/go-bindata/go-bindata
Running tests for APIVersion: v1,admissionregistration.k8s.io/v1,admissionregistration.k8s.io/v1beta1,admission.k8s.io/v1,admission.k8s.io/v1beta1,apps/v1,apps/v1beta1,apps/v1beta2,auditregistration.k8s.io/v1alpha1,authentication.k8s.io/v1,authentication.k8s.io/v1beta1,authorization.k8s.io/v1,authorization.k8s.io/v1beta1,autoscaling/v1,autoscaling/v2beta1,autoscaling/v2beta2,batch/v1,batch/v1beta1,batch/v2alpha1,certificates.k8s.io/v1beta1,coordination.k8s.io/v1beta1,coordination.k8s.io/v1,discovery.k8s.io/v1alpha1,extensions/v1beta1,events.k8s.io/v1beta1,imagepolicy.k8s.io/v1alpha1,networking.k8s.io/v1,networking.k8s.io/v1beta1,node.k8s.io/v1alpha1,node.k8s.io/v1beta1,policy/v1beta1,rbac.authorization.k8s.io/v1,rbac.authorization.k8s.io/v1beta1,rbac.authorization.k8s.io/v1alpha1,scheduling.k8s.io/v1alpha1,scheduling.k8s.io/v1beta1,scheduling.k8s.io/v1,settings.k8s.io/v1alpha1,storage.k8s.io/v1beta1,storage.k8s.io/v1,storage.k8s.io/v1alpha1,
+++ [1108 17:08:08] Running tests without code coverage
ok  	k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/describe/versioned	0.336s
```

**Comment by jackkleeman** [comment]
Ha, the issue only presents when you merge in master. Fixed

**Review by soltysh — APPROVED** [review]
The change seems reasonable, please squash the changes and I'll tag them for merge.

**Comment by jackkleeman** [comment]
Done @soltysh, thanks.

**Comment by jackkleeman** [comment]
Google is down :(

**Review by cmluciano — COMMENTED** [review]
TBH I'm still a little confused on if this change is right. I understand that the tests are showing that  specified policy itself does not include some explicit PolicyTypes therefore it makes sense that this policy should not assume that it is isolated for unspecified PolicyTypes.

My concern is that many CNI providers DO isolate for everyone if just one policy exists for Egress or Ingress. This means that an individual spec's describe may not show the whole picture with regards to cluster isolation. 

Does this make sense @danwinship @dcbw ?

File: staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go [hunk 3]
```diff
@@ -3745,13 +3745,38 @@ func describeNetworkPolicySpec(nps networkingv1.NetworkPolicySpec, w PrefixWrite
...
-	printNetworkPolicySpecIngressFrom(nps.Ingress, "    ", w)
-	w.Write(LEVEL_1, "Allowing egress traffic:\n")
-	printNetworkPolicySpecEgressTo(nps.Egress, "    ", w)
+
+	ingressEnabled, egressEnabled := getPolicyType(nps)
+	if ingressEnabled {
+		w.Write(LEVEL_1, "Allowing ingress traffic:\n")
+		printNetworkPolicySpecIngressFrom(nps.Ingress, "    ", w)
+	} else {
+		w.Write(LEVEL_1, "Allowing all ingress traffic\n")
```
Comment (danwinship) [inline staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go]: That's not right; it should say "Doesn't affect ingress traffic" or something like that. If ingress traffic would have been allowed without this policy, then it's still allowed with this policy, and if it would have been denied without this policy, then it's still denied with this policy. (Likewise for egress.)

**Review by danwinship — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go (hunk 3 above)
Comment (jackkleeman) [reply staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go]: Yeah I agree with that, will fix now

**Comment by jackkleeman** [comment]
@danwinship @cmluciano fair point - although technically in that case the statement `pods are isolated for egress connectivity` is also incorrect because another policy may allow it. There's an argument that these messages should just describe the policy's impact and not the potential sum of impacts of other policies. That said, I agree its a bit clearer, so I've pushed the change

**Comment by danwinship** [comment]
> although technically in that case the statement `pods are isolated for egress connectivity` is also incorrect because another policy may allow it.

*Technically* it is correct, because in the context of NetworkPolicy, pods are either "isolated" (meaning at least one NetworkPolicy targets them and so they only allow connections allowed by some NetworkPolicy) or "not isolated" (meaning no NetworkPolicies target them, and so they allow all connections). A pod that is targeted by an "allow all" policy is still considered "isolated"...

But yeah, the vast majority of people reading "kubectl describe" output aren't going to understand that and it would be good to come up with some better wording there. But even without that, this PR still improves things.

**Review by danwinship — APPROVED** [review]

**Comment by jackkleeman** [comment]
> A pod that is targeted by an "allow all" policy is still considered "isolated"
Yep, that interpretation makes a lot of sense 👍

**Comment by soltysh** [comment]
> But yeah, the vast majority of people reading "kubectl describe" output aren't going to understand that and it would be good to come up with some better wording there. But even without that, this PR still improves things.

Let's land this and we can improve further as we go, in that case.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
/priority important-longterm
